File: ios/bitchat/Views/Sonar/SonarWalletActivityScreen.swift
//
// SonarWalletActivityScreen.swift
// bitchat
//
// Wallet activity screen: gold-themed balance card, send/receive quick
// actions, transaction history from SonarPaymentActivityLedger. Ported from
// the Compose Multiplatform SonarWalletActivityScreen.
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

import SwiftUI

struct SonarWalletActivityScreen: View {
    @EnvironmentObject private var store: SonarAppStore

    @State private var toast: String?

    private var balanceSats: Int64 { store.balanceSats ?? 0 }
    private var entries: [SonarPaymentActivity] { store.paymentActivities }

    var body: some View {
        VStack(spacing: 0) {
            SNNavHeader(hairline: false, onBack: { store.pop() }) {
                SNHeaderName("Wallet")
            }

            ScrollView {
                VStack(spacing: 0) {
                    // ── Balance card ──
                    VStack(spacing: 0) {
                        SNIcon(name: .coin, size: 32)
                            .foregroundColor(SonarTheme.goldDeep)
                        HStack(alignment: .bottom, spacing: 6) {
                            Text(verbatim: sonarFormatSats(balanceSats))
                                .font(SonarTheme.uiFont(size: 36, weight: .heavy))
                                .foregroundColor(SonarTheme.text)
                            Text("sats")
                                .font(SonarTheme.uiFont(size: 15, weight: .bold))
                                .foregroundColor(SonarTheme.text3)
                                .padding(.bottom, 5)
                        }
                        .padding(.top, 10)
                        if let fiat = store.fiatText(balanceSats) {
                            Text(verbatim: fiat)
                                .font(SonarTheme.uiFont(size: 14))
                                .foregroundColor(SonarTheme.text2)
                                .padding(.top, 4)
                        }
                        Text("Lightning wallet")
                            .font(SonarTheme.uiFont(size: 12))
                            .foregroundColor(SonarTheme.text3)
                            .padding(.top, 6)
                    }
                    .frame(maxWidth: .infinity)
                    .padding(20)
                    .background(
                        RoundedRectangle(cornerRadius: 20, style: .continuous)
                            .fill(SonarTheme.goldSoft)
                    )
                    .padding(EdgeInsets(top: 10, leading: 14, bottom: 4, trailing: 14))

                    // ── Quick actions ──
                    HStack(spacing: 8) {
                        Button {
                            showToast("Open a chat to send or receive bitcoin")
                        } label: {
                            HStack(spacing: 6) {
                                SNIcon(name: .bolt, size: 18)
                                Text("Send")
                                    .font(SonarTheme.uiFont(size: 15, weight: .bold))
                            }
                            .foregroundColor(SonarTheme.onGold)
                            .frame(maxWidth: .infinity)
                            .frame(height: 48)
                            .background(Capsule().fill(SonarTheme.goldFill))
                        }
                        .buttonStyle(SNScaleStyle(scale: 0.97))

                        Button {
                            showToast("Open a chat to send or receive bitcoin")
                        } label: {
                            HStack(spacing: 6) {
                                SNIcon(name: .coin, size: 18)
                                Text("Receive")
                                    .font(SonarTheme.uiFont(size: 15, weight: .bold))
                            }
                            .foregroundColor(SonarTheme.text)
                            .frame(maxWidth: .infinity)
                            .frame(height: 48)
                            .background(Capsule().fill(SonarTheme.surface))
                        }
                        .buttonStyle(SNScaleStyle(scale: 0.97))
                    }
                    .padding(EdgeInsets(top: 4, leading: 14, bottom: 0, trailing: 14))

                    // ── Activity ──
                    SNSectionLabel("Activity")

                    if entries.isEmpty {
                        SNEmptyState(
                            icon: .bolt, iconSize: 24,
                            title: "No activity yet",
                            desc: "Send or receive bitcoin in any chat to see your transaction history here."
                        )
                        .frame(height: 200)
                    } else {
                        VStack(spacing: 0) {
                            ForEach(entries) { entry in
                                activityRow(entry)
                            }
                        }
                    }

                    Color.clear.frame(height: 40)
                }
            }
        }
        .background(SonarTheme.bg.ignoresSafeArea())
        .overlay(alignment: .bottom) { toastView }
        .animation(.easeOut(duration: 0.2), value: toast)
    }

    @ViewBuilder
    private var toastView: some View {
        if let toast {
            Text(verbatim: toast)
                .font(SonarTheme.uiFont(size: 13.5, weight: .medium))
                .foregroundColor(SonarTheme.text)
                .multilineTextAlignment(.center)
                .padding(EdgeInsets(top: 11, leading: 16, bottom: 11, trailing: 16))
                .background(
                    RoundedRectangle(cornerRadius: 13, style: .continuous)
                        .fill(SonarTheme.surface2)
                        .shadow(color: Color.black.opacity(0.18), radius: 12, y: 6)
                )
                .padding(.horizontal, 24)
                .padding(.bottom, 88)
                .transition(.move(edge: .bottom).combined(with: .opacity))
        }
    }

    private func showToast(_ text: String) {
        toast = text
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.6) {
            if toast == text { toast = nil }
        }
    }

    private func activityRow(_ entry: SonarPaymentActivity) -> some View {
        let sent = entry.direction == .outgoing
        let icon: SNIconName = sent ? .bolt : .coin

        let statusLabel: String = {
            switch entry.status {
            case .paid: return "Completed"
            case .pending: return "Pending"
            case .failed: return "Failed"
            }
        }()

        let statusColor: Color = {
            switch entry.status {
            case .paid: return SonarTheme.green
            case .pending: return SonarTheme.goldDeep
            case .failed: return SonarTheme.danger
            }
        }()

        let amountPrefix = sent ? "" : "+"
        let amountColor = sent ? SonarTheme.text : SonarTheme.green
        let fiat = store.fiatText(entry.sats)

        return HStack(spacing: 12) {
            RoundedRectangle(cornerRadius: 11, style: .continuous)
                .fill(SonarTheme.goldSoft)
                .frame(width: 38, height: 38)
                .overlay(
                    SNIcon(name: icon, size: 18)
                        .foregroundColor(SonarTheme.goldDeep)
                )

            VStack(alignment: .leading, spacing: 2) {
                Text(verbatim: sent ? "Sent" : "Received")
                    .font(SonarTheme.uiFont(size: 15.5, weight: .semibold))
                    .foregroundColor(SonarTheme.text)
                HStack(spacing: 0) {
                    Text(verbatim: statusLabel)
                        .font(SonarTheme.uiFont(size: 12.5))
                        .foregroundColor(statusColor)
                    Text(verbatim: " · Lightning")
                        .font(SonarTheme.uiFont(size: 12.5))
                        .foregroundColor(SonarTheme.text3)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            VStack(alignment: .trailing, spacing: 2) {
                Text(verbatim: "\(amountPrefix)\(sonarFormatSats(entry.sats)) sats")
                    .font(SonarTheme.uiFont(size: 16, weight: .semibold))
                    .foregroundColor(amountColor)
                if let fiat {
                    Text(verbatim: fiat)
                        .font(SonarTheme.uiFont(size: 12))
                        .foregroundColor(SonarTheme.text3)
                }
            }
        }
        .padding(EdgeInsets(top: 9, leading: 14, bottom: 9, trailing: 14))
    }
}

File: apps/sonar/composeApp/src/commonMain/kotlin/chat/bitchat/sonar/screens/SonarWalletActivityScreen.kt
package chat.bitchat.sonar.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import chat.bitchat.sonar.SonarAppState
import chat.bitchat.sonar.ToastBar
import chat.bitchat.sonar.payFmt
import chat.bitchat.sonar.wallet.SonarPaymentActivity
import chat.bitchat.sonar.wallet.WalletActivityItem
import chat.bitchat.sonar.wallet.WalletState
import chat.bitchat.sonar.ui.SNEmptyState
import chat.bitchat.sonar.ui.SNIcon
import chat.bitchat.sonar.ui.SNIconName
import chat.bitchat.sonar.ui.SNNavHeader
import chat.bitchat.sonar.ui.SNSectionLabel
import chat.bitchat.sonar.ui.sonar

@Composable
fun SonarWalletActivityScreen(state: SonarAppState) {
    val s = sonar
    // Subscribe to ledger changes so the list recomposes on new entries:
    // chat ⚡PAY receipts (payVersion) and direct wallet payment activity
    // (paymentActivityVersion) — both read through state, never the store.
    state.payVersion
    state.paymentActivityVersion

    val balanceSats = state.walletBalanceSats()
    // iOS `hasLiveRate` rule: the fiat subline renders ONLY when a live rate
    // exists — state.fiatOrNull already returns null otherwise (the
    // Money.formatFiat live-rate predicate), from the same state.rate snapshot
    // the rest of this screen uses.
    val fiat = state.fiatOrNull(balanceSats)
    // Chat receipts + direct wallet activity, deduped, newest first.
    val entries = state.walletActivity()

    Column(Modifier.fillMaxSize().background(s.bg)) {
        SNNavHeader("Wallet", hairline = false, onBack = { state.back() })

        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState())
                .padding(bottom = 40.dp)
        ) {
            // ── Balance card ──
            Column(
                Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 10.dp)
                    .clip(RoundedCornerShape(20.dp)).background(s.goldSoft)
                    .padding(20.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                SNIcon(SNIconName.Coin, 32.dp, s.goldDeep)
                Spacer(Modifier.height(10.dp))
                // pay-big: 42/800 amount with a small 15/700 text3 unit.
                Row(verticalAlignment = Alignment.Bottom) {
                    Text(
                        payFmt(balanceSats),
                        color = s.text,
                        fontSize = 42.sp,
                        fontWeight = FontWeight.ExtraBold,
                        letterSpacing = (-0.5).sp,
                    )
                    Spacer(Modifier.width(7.dp))
                    Text(
                        "sats",
                        color = s.text3,
                        fontSize = 15.sp,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.padding(bottom = 7.dp),
                    )
                }
                if (fiat != null) {
                    Spacer(Modifier.height(4.dp))
                    Text(fiat, color = s.text2, fontSize = 14.sp)
                }
                Spacer(Modifier.height(6.dp))
                Text(
                    "Lightning wallet",
                    color = s.text3,
                    fontSize = 12.sp,
                )
            }

            // ── Quick actions ──
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 4.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                // Send button
                Box(
                    Modifier.weight(1f).height(48.dp)
                        .clip(RoundedCornerShape(999.dp)).background(s.goldFill)
                        .clickable { state.toast = "Open a chat to send or receive bitcoin" },
                    contentAlignment = Alignment.Center,
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        SNIcon(SNIconName.Bolt, 18.dp, s.onGold)
                        Spacer(Modifier.width(6.dp))
                        Text(
                            "Send",
                            color = s.onGold,
                            fontSize = 15.sp,
                            fontWeight = FontWeight.Bold,
                        )
                    }
                }
                // Receive button
                Box(
                    Modifier.weight(1f).height(48.dp)
                        .clip(RoundedCornerShape(999.dp)).background(s.surface)
                        .clickable { state.toast = "Open a chat to send or receive bitcoin" },
                    contentAlignment = Alignment.Center,
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        SNIcon(SNIconName.Coin, 18.dp, s.text)
                        Spacer(Modifier.width(6.dp))
                        Text(
                            "Receive",
                            color = s.text,
                            fontSize = 15.sp,
                            fontWeight = FontWeight.Bold,
                        )
                    }
                }
            }

            // ── Activity section ──
            SNSectionLabel("Activity")

            if (entries.isEmpty()) {
                SNEmptyState(
                    icon = SNIconName.Activity,
                    title = "No activity yet",
                    desc = "Send or receive bitcoin in any chat to see your transaction history here.",
                )
            } else {
                Column(Modifier.fillMaxWidth()) {
                    entries.forEach { entry ->
                        ActivityRow(entry, state)
                    }
                }
            }
        }
    }

    state.toast?.let { ToastBar(it) { state.toast = null } }
}

@Composable
private fun ActivityRow(entry: WalletActivityItem, state: SonarAppState) {
    val s = sonar
    val sent = entry.sent
    val icon = if (sent) SNIconName.Bolt else SNIconName.Coin

    // iOS SonarWalletActivityScreen.activityRow: Completed/Pending/Failed.
    val statusLabel = when (entry.status) {
        SonarPaymentActivity.Status.Paid -> "Completed"
        SonarPaymentActivity.Status.Pending -> "Pending"
        SonarPaymentActivity.Status.Failed -> "Failed"
    }
    val statusColor = when (entry.status) {
        SonarPaymentActivity.Status.Paid -> s.green
        SonarPaymentActivity.Status.Pending -> s.goldDeep
        SonarPaymentActivity.Status.Failed -> s.danger
    }

    val amountPrefix = if (sent) "" else "+"
    val amountColor = if (sent) s.text else s.green
    val fiat = state.fiatOrNull(entry.sats)

    Row(
        Modifier.fillMaxWidth().padding(horizontal = 14.dp, vertical = 9.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // Icon tile
        Box(
            Modifier.size(38.dp).clip(RoundedCornerShape(11.dp)).background(s.goldSoft),
            contentAlignment = Alignment.Center,
        ) {
            SNIcon(icon, 18.dp, s.goldDeep)
        }
        Spacer(Modifier.width(12.dp))

        // Title + status
        Column(Modifier.weight(1f)) {
            Text(
                if (sent) "Sent" else "Received",
                color = s.text,
                fontSize = 15.5.sp,
                fontWeight = FontWeight.SemiBold,
            )
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    statusLabel,
                    color = statusColor,
                    fontSize = 12.5.sp,
                )
                Text(
                    " · Lightning",
                    color = s.text3,
                    fontSize = 12.5.sp,
                )
            }
        }

        // Amount + fiat
        Column(horizontalAlignment = Alignment.End) {
            Text(
                "$amountPrefix${payFmt(entry.sats)} sats",
                color = amountColor,
                fontSize = 16.sp,
                fontWeight = FontWeight.SemiBold,
            )
            if (fiat != null) {
                Text(
                    fiat,
                    color = s.text3,
                    fontSize = 12.sp,
                )
            }
        }
    }
}
